### PR TITLE
feat: fetch of edge version with a fallback option

### LINKED ISSUE #343: Display Edge Version in UI PWA
It would be useful for debugging and information to quickly check in the UI the version of the connected Ambianic Edge install.

The edge can expose a Rest endpoint that shares version and other vital metrics essential to diagnosing and solving common problems.

### PR DESCRIPTION
**Pull request by vickywane**
This PR resolves [this](https://github.com/ambianic/ambianic-ui/issues/343)

## Purpose
With the new modification in this PR, the connected edge device version is fetched and stored in the VueX store. For backward compatibility, a fallback has been implemented for this feature, considering only updated edge devices have the edge version in the `api/status` endpoint.

## Merge Checklist
- [x]  The code change is tested and works locally.
- [x]  The user and dev documentation is up to date.
- [x]  There is no commented-out code in this PR.
- [x]  No lint errors (use flake8)
- [x]  100% test coverage

### CODE REVIEW

**Comment by ivelin** [comment]
@vickywane I see this PR is converted as READY for review but I do not see a Review Request. Are you still working on it?

**Comment by vickywane** [comment]
Yes i am working on it.

File: src/store/index.js [hunk 1]
```diff
@@ -5,19 +5,34 @@ import { INITIALIZE_PNP } from './action-types.js'
 import { UPDATE_AVAILABLE } from './mutation-types'
 
 Vue.use(Vuex)
+let edgeDeviceVersion = require('@/../package.json').version
+
+try {
+  fetch('http://localhost:8778/api/status')
```
Comment (ivelin) [inline src/store/index.js:11]: @vickywane this is still a problem, as I mentioned in another comment. It is also what causes the Lighthouse fail.

Your are using the regular HTML 5 fetch() API instead of the p2p fetch over webrtc. Look closer into the mechanics of http access over webrtc.

Screen Shot 2021-03-12 at 2 51 32 PM

**Review by ivelin — CHANGES_REQUESTED** [review]

File: src/store/index.js (hunk 1 above)
Comment (ivelin) [reply src/store/index.js:11]: Screen Shot 2021-03-12 at 2 57 09 PM

File: jest-coverage/coverage-final.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
+{"/home/iamnwani/Desktop/OSS/ambianic-ui/src/App.vue": {"path":"/home/iamnwani/Desktop/OSS/ambianic-ui/src/App.vue","statementMap":{"0":{"start":{"line":19,"column":0},"end":{"line":19,"column":null}}},"fnMap":{},"branchMap":{},"s":{"0":1},"f":{},"b":{}}
```
Comment (ivelin) [inline jest-coverage/coverage-final.json:1]: @vickywane this coverage dump should be in .gitignore

File: src/store/index.js [hunk 3]
```diff
@@ -1,31 +1,40 @@
...
+      commit(EDGE_DEVICE_DETAILS, details)
+    }
   },
   modules: {
     pnp: PNPStore
   }
-})
+}
+
+const store = new Vuex.Store(applicationStore)
```
Comment (ivelin) [inline src/store/index.js:33]: @vickywane can you please elaborate on the semantic use of applixationStore in addition to the Vuex provided store structure? Why do we need to export both in this file?

**Review by ivelin — CHANGES_REQUESTED** [review]

File: src/store/index.js (hunk 3 above)
Comment (vickywane) [reply src/store/index.js:33]: @ivelin i exported the applicationStore so i can isolate and test it when testing components connected it. Being the default store, i couldnt write tests for it like the `pnp` store that’s exported and used as a Vuex module. 

See an example usage [here](https://github.com/ambianic/ambianic-ui/pull/582/files#diff-a0e2b31817857d0241e7b5084937ab2ba1bfe4c411405ec24d6ba64a392b3e9f)

File: jest-coverage/coverage-final.json (hunk 2 above)
Comment (vickywane) [reply jest-coverage/coverage-final.json:1]: I would add it.

Comment (vickywane) [reply jest-coverage/coverage-final.json:1]: I have added the `jest-coverage` directory to the `.gitignore`

**Comment by ivelin** [comment]
@vickywane this PR has a merge conflict after the merge of the navbar icon spacing PR. Can you please sync with main and resubmit for review.

**Review by ivelin — CHANGES_REQUESTED** [review]
@vickywane please resolve merge conflicts and resubmit.

File: src/store/index.js (hunk 3 above)
Comment (ivelin) [reply src/store/index.js:33]: @vickywane Let's move the new code to a separate module. Let's put EDGE_DEVICE_DETAILS and FETCH_EDGE_DEVICE_DETAILS in an edgeDevice module in the Vuex store instead of the default top level store.

**Review by ivelin — CHANGES_REQUESTED** [review]

File: src/store/index.js [hunk 4]
```diff
@@ -1,40 +1,32 @@
...
   actions: {
-    [FETCH_EDGE_DEVICE_DETAILS] ({ commit }, details) {
-      console.log('I AM CALLED')
-      commit(EDGE_DEVICE_DETAILS, details)
-    }
   },
   modules: {
-    pnp: PNPStore
+    pnp: PNPStore,
+    edgeDeviceStore: EdgeDeviceStore
```
Comment (ivelin) [inline src/store/index.js:23]: @vickywane the use of Store suffix in the name is redundant and inconsistent with the naming of the other store modules. Let's rename `edgeDeviceStore` to `edgeDevice`.

File: tests/unit/components/about.spec.js [hunk 5]
```diff
@@ -4,8 +4,9 @@ import Vuetify from 'vuetify'
 import VueX from 'vuex'
 import { cloneDeep } from 'lodash'
 import About from '@/views/About.vue'
-import { PEER_CONNECTED } from '../../../src/store/mutation-types'
-import { applicationStore } from '@/store/index.js'
+import { PEER_CONNECTED } from '@/store/mutation-types'
+import EdgeDeviceStore from '@/store/edge-device.js'
```
Comment (ivelin) [inline tests/unit/components/about.spec.js:8]: @vickywane This should be triggering an error. `edge-device.js` exports `edgeDeviceModule` but this line imports `EdgeDeviceStore`. What gives?

**Review by ivelin — CHANGES_REQUESTED** [review]

File: tests/unit/components/about.spec.js (hunk 5 above)
Comment (vickywane) [reply tests/unit/components/about.spec.js:8]: It wouldn't trigger an error as it's a **default export** from the `edge-device.js` file, and not a **named export**. 

I however understand that it's inconsistent and I'll fix that.

File: src/store/index.js (hunk 4 above)
Comment (vickywane) [reply src/store/index.js:23]: I will make this change and revert

File: cypress/integration/ambianic-tests/settings.spec.js [hunk 6]
```diff
@@ -34,7 +34,11 @@ context('Settings', () => {
     it("It display connected edge version after connection", () => {
       cy.get('[data-cy=sendRemotePeerID]').click()
 
-      const versionNumber = require('../../../package.json').version
-      cy.get('[data-cy=title-text]').should('contain.text', versionNumber )
+      cy.window().then(win => {
+        win.__store__.commit('PEER_CONNECTED', null)
+
+        const versionNumber = require('../../../package.json').version
```
Comment (ivelin) [inline cypress/integration/ambianic-tests/settings.spec.js:40]: @vickywane what is the purpose of this test? From what I can tell, it checks that the page shows the PWA version number, not the edge device version.

File: tests/unit/components/settings/settings.spec.js [hunk 7]
```diff
@@ -58,7 +58,7 @@ describe('NavBar', () => {
   test('Connected Edge device version is shown', () => {
     const localEdgeVersion = require('@/../package.json').version
```
Comment (ivelin) [inline tests/unit/components/settings/settings.spec.js:59]: @vickywane package.json holds the PWA version, not the edge device version.

File: tests/unit/components/settings/settings.spec.js [hunk 8]
```diff
@@ -58,7 +58,7 @@ describe('NavBar', () => {
...
 
-    store.state.version = localEdgeVersion
+    store.state.edgeDevice.version = localEdgeVersion
```
Comment (ivelin) [inline tests/unit/components/settings/settings.spec.js:61]: @vickywane this comparison is meaningless. Compares edgeDeviceVersion to PWA version.

File: src/store/action-types.js [hunk 9]
```diff
@@ -11,3 +11,4 @@ export const PEER_AUTHENTICATE = 'peerAuthenticate'
 export const REMOVE_REMOTE_PEER_ID = 'removeRemotePeerId'
 export const CHANGE_REMOTE_PEER_ID = 'newRemotePeerId'
 export const HANDLE_PEER_CONNECTION_ERROR = 'handlePeerConnectionError'
+export const FETCH_EDGE_DEVICE_DETAILS = 'FETCH_EDGE_DEVICE_DETAILS'
```
Comment (ivelin) [inline src/store/action-types.js:14]: @vickywane please follow the naming convention for other actions. The string value corresponds to a JS function name so it should be camel cased.

File: src/store/index.js [hunk 10]
```diff
@@ -1,15 +1,15 @@
 import Vue from 'vue'
 import Vuex from 'vuex'
 import PNPStore from './pnp.js'
```
Comment (ivelin) [inline src/store/index.js:3]: @vickywane please sync with upstream.

File: src/views/About.vue [hunk 11]
```diff
@@ -72,16 +69,47 @@
...
+import { EdgeAPI } from '../remote/edgeAPI'
 
 export default {
   components: {
     AmbListItem
   },
   computed: {
     ...mapState({
-      version: state => state.version
+      version: (state) => state.edgeDevice.version,
```
Comment (ivelin) [inline src/views/About.vue:83]: @vickywane this replaces the PWA version display with edge device version display. Not a good UX design.
The About page should continue to show the PWA version as before. Maybe we should add a clarification that this is the UI app version so its clear. Edge device version should be shown only next to a device details section, which is currently in the Settings.vue page.

File: src/views/About.vue [hunk 12]
```diff
@@ -40,7 +37,7 @@
         >
           <v-flex>
             <amb-list-item
-              id="version-info"
+              id="version-element"
               :title="version"
               icon-name="alpha-v-circle-outline"
               subtitle="Release Version"
```
Comment (ivelin) [inline src/views/About.vue:43]: @vickywane let's update the value from "Release Version" to "UI App Version" so it's clear what the version number relates to.

File: src/views/Settings.vue [hunk 13]
```diff
@@ -61,6 +61,7 @@
                   />
                   <amb-list-item
                     :title="version"
+                    id="version-element"
                     subtitle="Release Version"
```
Comment (ivelin) [inline src/views/Settings.vue:65]: @vickywane let's update the text from "Release Version" to "Edge Software Version"

File: src/views/Settings.vue [hunk 14]
```diff
@@ -329,7 +345,8 @@ export default {
         state.pnp.peerConnectionStatus === PEER_CONNECTED,
       edgePeerId: state => state.pnp.remotePeerId,
       peerFetch: state => state.pnp.peerFetch,
-      version: state => state.version
+      version: state => state.edgeDevice.version,
```
Comment (ivelin) [inline src/views/Settings.vue:348]: @vickywane I can see how the existing code confused you. It incorrectly showed the UI App version on the Edge Device card. The new code that you committed here fixes that by replacing `state.version` with `state.edgeDevice.version`. Let's update the corresponding description label next to the version number to clearly state that this is the edge device version.

File: tests/unit/components/about.spec.js [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+      store
+    })
+  })
+
+  afterEach(() => {
+    wrapper.destroy()
+  })
+
+  test('edgeDevice details is retrieved from store', async () => {
+    const localEdgeVersion = require('@/../package.json').version
```
Comment (ivelin) [inline tests/unit/components/about.spec.js:42]: @vickywane this test is incorrect. package.json holds the PWA version, not the edge device version.

File: tests/unit/components/about.spec.js [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+import { mount, createLocalVue } from '@vue/test-utils'
+import Vuetify from 'vuetify'
+import VueX from 'vuex'
+import { cloneDeep } from 'lodash'
+import About from '@/views/About.vue'
+import { PEER_CONNECTED } from '@/store/mutation-types'
+import edgeDevice from '@/store/edge-device.js'
+import pnpStoreModule from '@/store/pnp'
+
+describe('About Page', () => {
```
Comment (ivelin) [inline tests/unit/components/about.spec.js:11]: @vickywane About page only shows the PWA UI app version, so tests should reflect that. No point in testing for edge device version because its not shown on the About page.

File: tests/unit/components/settings/settings.spec.js [hunk 17]
```diff
@@ -63,4 +54,21 @@ describe('NavBar', () => {
...
+    store.state.pnp.peerConnectionStatus = PEER_CONNECTED
+    store.state.pnp.remotePeerId = '1234-1234-1234-1234-1234'
+
+    const component = mount(Settings, {
+      localVue,
+      vuetify,
+      store
+    })
+
+    const versionElement = component.get('#version-element')
```
Comment (ivelin) [inline tests/unit/components/settings/settings.spec.js:71]: @vickywane this test needs to be rewritten. It should mock an edge device status API fetch that returns and edge version. Then compare the edge device version displayed on the page to the version from the status API response.

**Review by ivelin — CHANGES_REQUESTED** [review]
@vickywane some cleanup required. See inline comments.

File: cypress/integration/ambianic-tests/settings.spec.js (hunk 6 above)
Comment (vickywane) [reply cypress/integration/ambianic-tests/settings.spec.js:40]: I had a previous thinking that the About and Settings page showed the version of the connected edge device.

File: tests/unit/components/settings/settings.spec.js (hunk 7 above)
Comment (vickywane) [reply tests/unit/components/settings/settings.spec.js:59]: I had previous thinking that the About and Settings page showed the version of the connected edge device. 

I will correct it this.

**Comment by vickywane** [comment]
@ivelin 

Following your clarification on the PWA and Edge Device versions, I have separated them by using the `version` in the `package.json` file without making any call to the Edge Device from the `About` component. 

The Setting page now uses the `version` fetched from the connected edge device, however, it still uses the `version` in the `package.json` as a fallback until the actual version is fetched. 
Is this something that should be changed? 

Also I put back the `version` state for the PWA in the `index.js` store, while the state for the Edge Device in the `edgeDevice.js` module. Should the naming of the two version state be adjusted to reflect their usage? 
E.g `appVersion` and `edgeDeviceVersion` ?

File: cypress/integration/ambianic-tests/settings.spec.js (hunk 6 above)
Comment (ivelin) [reply cypress/integration/ambianic-tests/settings.spec.js:40]: Yes, I realize the previous implementation was confusing. We will correct that with this PR.

Please update this test to check for edge device version in the settings page.

File: src/store/action-types.js (hunk 9 above)
Comment (ivelin) [reply src/store/action-types.js:14]: @vickywane did you see this comment?

File: src/store/edge-device.js [hunk 18]
```diff
@@ -0,0 +1,29 @@
+import { FETCH_EDGE_DEVICE_DETAILS } from './action-types.js'
+import { EDGE_DEVICE_DETAILS } from './mutation-types'
+
+const state = {
+  version: require('@/../package.json').version
```
Comment (ivelin) [inline src/store/edge-device.js:5]: @vickywane version here should not default to the PWA version. It should be undefined until the PWA connects to an edgeDevice and fetches its version.

File: tests/unit/components/about.spec.js [hunk 19]
```diff
@@ -0,0 +1,52 @@
+import Vue from 'vue'
+import { mount, createLocalVue } from '@vue/test-utils'
+import Vuetify from 'vuetify'
+import VueX from 'vuex'
+import About from '@/views/About.vue'
+import { pnpStoreModule } from '../../../src/store/pnp'
+
+const localEdgeVersion = require('@/../package.json').version
```
Comment (ivelin) [inline tests/unit/components/about.spec.js:8]: @vickywane there is no edge device version shown on the About page. Only the PWA version. Please correct variable names and logic.

File: tests/unit/components/settings/settings.spec.js [hunk 20]
```diff
@@ -26,10 +30,12 @@ describe('NavBar', () => {
 
   beforeEach(() => {
     state = {
-      pnp: {
-        peerConnectionStatus: jest.fn(),
-        edgePeerId: jest.fn()
-      }
+      pnp: cloneDeep(pnpStoreModule),
+      // pnp: {
```
Comment (ivelin) [inline tests/unit/components/settings/settings.spec.js:34]: @vickywane do we need to commit code that's commented out. Can you please clean up.

File: tests/unit/components/settings/settings.spec.js [hunk 21]
```diff
@@ -75,4 +81,29 @@ describe('NavBar', () => {
...
+
+    const component = mount(Settings, {
+      localVue,
+      vuetify,
+      router,
+      store: newStore
+    })
+
+    const versionElement = component.get('#version-element')
+    // expect(versionElement.find('#title').text()).toBe(localEdgeVersion)
```
Comment (ivelin) [inline tests/unit/components/settings/settings.spec.js:107]: @vickywane where is the actual test assertion? This line is just commented out.

File: src/store/edge-device.js [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+const mutations = {
+  [EDGE_DEVICE_DETAILS] (state, edgeDetails) {
+    state.version = edgeDetails.version
+  }
+}
+
+const actions = {
+  [FETCH_EDGE_DEVICE_DETAILS] ({ commit }, details) {
```
Comment (ivelin) [inline src/store/edge-device.js:15]: @vickywane this action is not very useful, because it only commits a single mutation. 
Your intention was probably to use it as a mechanism to call the edgeAPI and subscribe to mutation notifications in the vue code. Please implement that or simply remove the action and only use the mutation in the vue  code.

**Review by ivelin — CHANGES_REQUESTED** [review]
Some cleanup still needed. See all unresolved comments.

File: src/store/action-types.js (hunk 9 above)
Comment (vickywane) [reply src/store/action-types.js:14]: Yes. 

I wasn't done with all changes at the time i made the last comment on this pull request.

File: src/store/edge-device.js (hunk 18 above)
Comment (vickywane) [reply src/store/edge-device.js:5]: I have made this correction. 

It now defaults to `null` until it's loaded. However, would it not be better to have a default loading text for UX?

File: tests/unit/components/about.spec.js (hunk 19 above)
Comment (vickywane) [reply tests/unit/components/about.spec.js:8]: I have updated the variable name and also stopped setting the state manually.

File: src/store/edge-device.js (hunk 22 above)
Comment (vickywane) [reply src/store/edge-device.js:15]: I have removed this action and called the commit method directly from the concerned component and test suite.

File: tests/unit/components/settings/settings.spec.js (hunk 21 above)
Comment (vickywane) [reply tests/unit/components/settings/settings.spec.js]: I have improved this test suite to assert that an edge version is shown.

File: tests/unit/components/settings/settings.spec.js (hunk 20 above)
Comment (vickywane) [reply tests/unit/components/settings/settings.spec.js:34]: I have removed the comment/

File: tests/unit/components/settings/settings.spec.js (hunk 17 above)
Comment (vickywane) [reply tests/unit/components/settings/settings.spec.js:106]: I'd like to object to this here. 

There is already a [test suite](https://github.com/vickywane/ambianic-ui/blob/update/edge-version/tests/unit/remote/edgeAPI.spec.js#L84) in the `edgeApi.spec.js` file that mocks a request to fetch the edge version.
 
Here, the assertion is on displaying the version in the store in the component.

File: jest-coverage/coverage-final.json (hunk 2 above)
Comment (ivelin) [reply jest-coverage/coverage-final.json:1]: @vickywane why is this file still in the PR?

**Review by ivelin — CHANGES_REQUESTED** [review]
Some further cleanup needed.

File: cypress/integration/ambianic-tests/settings.spec.js [hunk 23]
```diff
@@ -30,4 +30,19 @@ context('Settings', () => {
...
+
+      cy.get('[data-cy=sendRemotePeerID]').click()
+
+      cy.window().then(win => {
+        win.__store__.commit('PEER_CONNECTED', null)
+        win.__store__.commit('EDGE_DEVICE_DETAILS', { 
+          version
+         })
+
+        cy.get('[data-cy=title-text]').should('contain.text', version )
```
Comment (ivelin) [inline cypress/integration/ambianic-tests/settings.spec.js:45]: @vickywane did you check whether this test fails if the assertion fails. I am not sure whether cypress will fail when the assertion is embedded in a promise or it will just always pass and ignore the result of the assertion. It seems to have different evaluation semantics than jest.
https://docs.cypress.io/api/commands/window#No-Args

**Review by ivelin — CHANGES_REQUESTED** [review]

File: cypress/integration/ambianic-tests/settings.spec.js (hunk 23 above)
Comment (vickywane) [reply cypress/integration/ambianic-tests/settings.spec.js:46]: Yes. 

I just tested this particular e2e test. It fails when the version text is not found.

Comment (ivelin) [reply cypress/integration/ambianic-tests/settings.spec.js:46]: OK. Thank you for verifying.

File: cypress-coverage/coverage-final.json [hunk 24]
```diff
@@ -0,0 +1,40 @@
...ne":33,"column":1}},"line":2},"1":{"name":"(anonymous_1)","decl":{"start":{"line":3,"column":13},"end":{"line":3,"column":14}},"loc":{"start":{"line":3,"column":19},"end":{"line":5,"column":3}},"line":3},"2":{"name":"(anonymous_2)","decl":{"start":{"line":7,"column":33},"end":{"line":7,"column":34}},"loc":{"start":{"line":7,"column":39},"end":{"line":13,"column":3}},"line":7},"3":{"name":"(anonymous_3)","decl":{"start":{"line":15,"column":30},"end":{"line":15,"column":31}},"loc":{"start":{"line":15,"column":36},"end":{"line":20,"column":3}},"line":15},"4":{"name":"(anonymous_4)","decl":{"start":{"line":22,"column":30},"end":{"line":22,"column":31}},"loc":{"start":{"line":22,"column":36},"end":{"line":27,"column":3}},"line":22},"5":{"name":"(anonymous_5)","decl":{"start":{"line":29,"column":27},"end":{"line":29,"column":28}},"loc":{"start":{"line":29,"column":33},"end":{"line":32,"column":3}},"line":29}},"branchMap":{},"s":{"0":1,"1":1,"2":4,"3":1,"4":1,"5":1,"6":1,"7":1,"8":1,"9":1,"10":1,"11":1,"12":1,"13":1,"14":1,"15":1},"f":{"0":1,"1":4,"2":1,"3":1,"4":1,"5":1},"b":{},"_coverageSchema":"1a1c01bbd47fc00a2c39e90264f33305004495a9","hash":"775d48c0deba75af90cbe529474f441e94d6f94e"}
```
[the remark this answers is not in the document]
Comment (ivelin) [reply cypress-coverage/coverage-final.json:1]: @vickywane you updated .gitignore but left the reports in the PR. Please cleanup.

**Review by ivelin — CHANGES_REQUESTED** [review]

**Comment by vickywane** [comment]
> @vickywane you updated .gitignore but left the reports in the PR. Please cleanup.

I have deleted the `cypress` and `jest` coverage reports.

File: src/views/Settings.vue [hunk 25]
```diff
@@ -318,6 +321,15 @@ export default {
     localEdgeAddress () {
       this.edgeAddress = undefined
       this.$store.dispatch(REMOVE_REMOTE_PEER_ID)
+    },
+    async fetchEdgeDetails () {
+      try {
+        const details = await this.edgeAPI.getEdgeStatus()
```
Comment (ivelin) [inline src/views/Settings.vue:327]: @vickywane I just realized that this code does not handle older Edge device versions. If the status API doesn't exist or is not available, this code will through undesirable exceptions.

File: src/store/edge-device.js [hunk 26]
```diff
@@ -0,0 +1,18 @@
+import { EDGE_DEVICE_DETAILS } from './mutation-types'
+
+const state = {
+  edgeSoftwareVersion: null
+}
+
+const mutations = {
+  [EDGE_DEVICE_DETAILS] (state, edgeDetails) {
+    state.edgeSoftwareVersion = edgeDetails.version
```
Comment (ivelin) [inline src/store/edge-device.js:9]: @vickywane we need some error handling for edge cases. If `edgeDetails` is undefined or doesn't have a fully populated object structure, `edgeDetails.version` will populate `state.edgeSoftwareVersion` with a value that is not user friendly. In such cases, let's make sure `state.edgeSoftwareVersion` is left undefined so that the UI components can recognize the issue and render a user friendly message.

File: src/views/Settings.vue [hunk 27]
```diff
@@ -62,7 +62,8 @@
                     />
                     <amb-list-item
                       :title="version"
```
Comment (ivelin) [inline src/views/Settings.vue:64]: @vickywane let's add some UX edge case handling here. If version is undefined due to some error accessing it via the API, we should show something like:
`Version information unavailable. Please make sure your edge device is online and up to date."

**Review by ivelin — CHANGES_REQUESTED** [review]
We need error handling for edge cases.

File: src/views/Settings.vue (hunk 27 above)
Comment (vickywane) [reply src/views/Settings.vue:64]: I have added this use case and I also added a test case for it also.

File: src/store/edge-device.js (hunk 26 above)
Comment (vickywane) [reply src/store/edge-device.js]: Alright, this makes sense.

**Comment by ivelin** [comment]
@vickywane did you notice that 3 test are failing in the CI?

**Comment by ivelin** [comment]
@vickywane the UI does not behave as expected when I test against an edge device that does not offer a status API. In fact, the error that I see in the Dev Console suggests that there is an issue independent of the status API. Please investigate thoroughly and add test coverage. See error message and screenshot below. To reproduce: Start the UI, got to About, then go to Settings. Edge Version is blank. 

BTW, at this point the log does not show any attempt to fetch the status API. Most likely because the `watch` callback for `isEdgeConnected` is never triggered, because the edge connection was established before the Settings page was loaded and its not changing any further. This watch doesn't seem to be accomplishing the goal of updating status when edge is connected. The code should take into account that the connection might have been established before visiting the settings page.

Anyway, to continue with reproducing this issue, reload the Settings page in the browser. The error below occurs.

```
Settings.vue?e12e:331 TypeError: Cannot read property 'getEdgeStatus' of undefined
    at _callee$ (Settings.vue?e12e:325)
    at tryCatch (runtime.js?96cf:62)
    at Generator.invoke [as _invoke] (runtime.js?96cf:288)
    at Generator.prototype.<computed> [as next] (runtime.js?96cf:114)
    at asyncGeneratorStep (asyncToGenerator.js?c964:3)
    at _next (asyncToGenerator.js?c964:25)
    at eval (asyncToGenerator.js?c964:32)
    at new Promise (<anonymous>)
    at eval (asyncToGenerator.js?c964:21)
    at VueComponent.fetchEdgeDetails (Settings.vue?e12e:325)
_callee$ @ Settings.vue?e12e:331
tryCatch @ runtime.js?96cf:62
invoke @ runtime.js?96cf:288
prototype.<computed> @ runtime.js?96cf:114
asyncGeneratorStep @ asyncToGenerator.js?c964:3
_next @ asyncToGenerator.js?c964:25
eval @ asyncToGenerator.js?c964:32
eval @ asyncToGenerator.js?c964:21
fetchEdgeDetails @ Settings.vue?e12e:325
_callee2$ @ Settings.vue?e12e:375
tryCatch @ runtime.js?96cf:62
invoke @ runtime.js?96cf:288
prototype.<computed> @ runtime.js?96cf:114
asyncGeneratorStep @ asyncToGenerator.js?c964:3
_next @ asyncToGenerator.js?c964:25
eval @ asyncToGenerator.js?c964:32
eval @ asyncToGenerator.js?c964:21
isEdgeConnected @ Settings.vue?e12e:375
invokeWithErrorHandling @ vue.runtime.esm.js?2b0e:1863
run @ vue.runtime.esm.js?2b0e:4584
flushSchedulerQueue @ vue.runtime.esm.js?2b0e:4326
eval @ vue.runtime.esm.js?2b0e:1989
flushCallbacks @ vue.runtime.esm.js?2b0e:1915
Promise.then (async)
timerFunc @ vue.runtime.esm.js?2b0e:1942
nextTick @ vue.runtime.esm.js?2b0e:1999
queueWatcher @ vue.runtime.esm.js?2b0e:4418
update @ vue.runtime.esm.js?2b0e:4560
notify @ vue.runtime.esm.js?2b0e:730
reactiveSetter @ vue.runtime.esm.js?2b0e:1055
eval @ pnp.js?d013:113
wrappedMutationHandler @ vuex.esm.js?2f62:844
commitIterator @ vuex.esm.js?2f62:466
eval @ vuex.esm.js?2f62:465
_withCommit @ vuex.esm.js?2f62:624
commit @ vuex.esm.js?2f62:464
boundCommit @ vuex.esm.js?2f62:409
_callee8$ @ pnp.js?d013:502
tryCatch @ runtime.js?96cf:62
invoke @ runtime.js?96cf:288
prototype.<computed> @ runtime.js?96cf:114
asyncGeneratorStep @ asyncToGenerator.js?c964:3
_next @ asyncToGenerator.js?c964:25
Promise.then (async)
asyncGeneratorStep @ asyncToGenerator.js?c964:13
_next @ asyncToGenerator.js?c964:25
eval @ asyncToGenerator.js?c964:32
eval @ asyncToGenerator.js?c964:21
eval @ pnp.js?d013:478
wrappedActionHandler @ vuex.esm.js?2f62:851
dispatch @ vuex.esm.js?2f62:516
boundDispatch @ vuex.esm.js?2f62:406
eval @ pnp.js?d013:273
setTimeout (async)
eval @ pnp.js?d013:273
i.emit @ peerjs.min.js?a0bc:46
dataChannel.onopen @ peerjs.min.js?a0bc:62
```
Screen Shot 2021-08-12 at 8 55 40 AM

**Comment by ivelin** [comment]
@vickywane please add more test cases for handling edge cases and graceful error handling when accessing edge devices that are offline or do not support the status API yet.

**Review by ivelin — CHANGES_REQUESTED** [review]

**Comment by ivelin** [comment]
@vickywane I tested manually this feature. Looks good. Let's wrap up the test coverage and merge.

**Comment by vickywane** [comment]
I have created a new test suite for the `edgeDevice` module that covers the lines you pointed out.

File: src/views/Settings.vue [hunk 28]
```diff
@@ -318,6 +327,19 @@ export default {
...
+    async fetchEdgeDetails () {
+      try {
+        const details = await this.edgeAPI.getEdgeStatus()
+
+        if (!details.version) {
+          this.edgeDeviceError = 'Edge version information is currently unavailable. Please make sure your edge device is online and up to date.'
+        }
+
+        await this.$store.commit(EDGE_DEVICE_DETAILS, details)
+      } catch (e) {
```
[the remark this answers is not in the document]
Comment (vickywane) [reply src/views/Settings.vue:344]: I have covered this line within the `edgeDevice.spec.js` file.

File: src/views/Settings.vue [hunk 29]
```diff
@@ -318,6 +327,19 @@ export default {
...
       this.edgeAddress = undefined
       this.$store.dispatch(REMOVE_REMOTE_PEER_ID)
+    },
+    async fetchEdgeDetails () {
+      try {
+        const details = await this.edgeAPI.getEdgeStatus()
+
+        if (!details.version) {
+          this.edgeDeviceError = 'Edge version information is currently unavailable. Please make sure your edge device is online and up to date.'
+        }
```
[the remark this answers is not in the document]
Comment (vickywane) [reply src/views/Settings.vue:341]: I have covered this line within the `edgeDevice.spec.js` file.

File: src/store/edge-device.js [hunk 30]
```diff
@@ -0,0 +1,22 @@
...
+  edgeSoftwareVersion: null
+}
+
+const mutations = {
+  [EDGE_DEVICE_DETAILS] (state, edgeDetails) {
+    console.log(edgeDetails, 'EDGE STORE \n \n \n \n \n')
+
+    if (edgeDetails.version) {
+      state.edgeSoftwareVersion = edgeDetails.version
+    }
```
[the remark this answers is not in the document]
Comment (vickywane) [reply src/store/edge-device.js:11]: I have covered this line within the `edgeDevice.spec.js` file.

**Review by ivelin — CHANGES_REQUESTED** [review]
@vickywane During testing tonight, I found a few issues that I didn't see before:

1. Error message was not tested. The error property was not declared in ListeItem.vue which caused problems. 
2. If the user visits the timeline page during which time connection to edge is established and then they go to the settings page, the `watch` for `isEdgeConnected` did not trigger and the edge version was never shown.
3. In the process of working on 1&2 I noticed that we've overcomplicated the markup in ListeView and simplified it using basic vuetify components.
4. `settings.spec.js` moved from `/tests/unit/components` to `/tests/unit/views` to mirror the `src/` file structure of the code these tests are intended to cover. Without a consistent test file location convention we will lose track which tests correspond to what part of the source code.
5. `edgeDevice.spec.js` only tests store mutations but pulls in the Settings page. Cleaned up to focus on testing the store without noisy context.

Please review my changes and comment.

Screen Shot 2021-08-16 at 10 39 15 PM
Screen Shot 2021-08-16 at 10 36 58 PM

**Comment by vickywane** [comment]
Thanks for the cleaning up! 

I just pulled your changes and tested them out locally. 

> Error message was not tested. The error property was not declared in ListeItem.vue which caused problems.
I checked and discovered the error property was removed when I merged the last PR that merged into master. 

> edgeDevice.spec.js only tests store mutations but pulls in the Settings page. Cleaned up to focus on testing the store without noisy context.
This is good. 

> settings.spec.js moved from /tests/unit/components to /tests/unit/views to mirror the src/ file structure of the code these tests are intended to cover. Without a consistent test file location convention we will lose track which tests correspond to what part of the source code.

I noticed this when I pulled the master branch. I also planned to call your attention as to why we had three test files for a single component.

**Comment by vickywane** [comment]
Overall, I think the objective of this PR has been achieved.  

Please merge if there's nothing else to be implemented.
